runtime/compiler/env/VMJ9Server.cpp

```diff
@@ -486,32 +486,77 @@ TR_J9ServerVM::classHasBeenReplaced(TR_OpaqueClassBlock *clazz)
    return ((classDepthAndFlags & J9AccClassHotSwappedOut) != 0);
    }
 
+bool
+TR_J9ServerVM::checkCHTableIfClassInfoExistsAndHasBeenExtended(TR_OpaqueClassBlock *clazz, bool &bClassHasBeenExtended)
+   {
+   JITServerPersistentCHTable *table = (JITServerPersistentCHTable*)(_compInfo->getPersistentInfo()->getPersistentCHTable());
+   TR_PersistentClassInfo *classInfo = table->findClassInfoAfterLocking(clazz, this);
+   if (classInfo)
+      {
+      bClassHasBeenExtended = classInfo->getFirstSubclass() ? true : false;
+      return true;
+      }
+   return false;
+   }
+
 bool
 TR_J9ServerVM::classHasBeenExtended(TR_OpaqueClassBlock *clazz)
    {
-   if(!clazz)
+   if (!clazz)
       return false;
-   uintptr_t classDepthAndFlags = 0;
+
+   ClientSessionData *clientSessionData = _compInfoPT->getClientData();
    JITServer::ServerStream *stream = _compInfoPT->getMethodBeingCompiled()->_stream;
-   bool dataFromCache = JITServerHelpers::getAndCacheRAMClassInfo((J9Class *)clazz, _compInfoPT->getClientData(), stream, JITServerHelpers::CLASSINFO_CLASS_DEPTH_AND_FLAGS, (void *)&classDepthAndFlags);
 
-   if(dataFromCache)
+   // Check the CHTable first and then the ROMClass cache.
+   // Checking the CHTable and checking the ROMClass cache both take monitors.
+   // They should not be nested into each other to avoid the potential
+   // deadlock when being accessed in different orders in other parts of the code.
+   bool bClassHasBeenExtended = false;
+   bool bIsClassInfoInCHTable = checkCHTableIfClassInfoExistsAndHasBeenExtended(clazz, bClassHasBeenExtended);
+
+   // The class data is in the CHTable and the flag is set.
+   if (bIsClassInfoInCHTable && bClassHasBeenExtended)
+      return true;
```

Comment:
If CHTable is always checked first here, I don't think there is a need to update the bit in the ROMClass cache. 
To be on the safe side, it still checks the ROMClass cache if the class info exists in the CHTable but doesn't have subclasses.